### ISSUE TITLE
Infer type hints for Deref-bound types

### ISSUE THREAD

**Issue by arik-so** [intellij-rust/intellij-rust#10444]
## Environment

* **IntelliJ Rust plugin version:** `0.4.194.5382-231`
* **Rust toolchain version:** `rustup 1.26.0 (5af9b9484 2023-04-05)`
* **IDE name and version:** `CLion 2023.1.2, Build #CL-231.8770.66, built on April 27, 2023`
* **Operating system:** `Ventura 13.3.1 (a) (22E772610a)`

## Problem description

When using a generic argument whose type is specified behind one layer of `Deref`, autocompletion and type inference don't work.

## Steps to reproduce

Consider this code snippet:

```rust
trait Foo {
    fn do_foo_stuff(&self) -> String;
}

fn bar<F: core::ops::Deref>(foo: &F) -> String where F::Target: Foo {
    let result = foo.do_foo_stuff();
    result
}
```

When just typing `foo.`, the autocompletion list does not include the `do_foo_stuff()` method. Additionally, the type of `result` is shown as Unknown, even though it is clearly `String`, which, in some scenarios, results in a spurious red error underlining.

<img width="650" alt="Screenshot 2023-05-08 at 10 37 22 AM" src="https://user-images.githubusercontent.com/927534/236892400-800934ec-f688-4afa-8141-cf6e66e38cde.png">
